### ISSUE TITLE
add city search functionality

### ISSUE THREAD

**Issue by choldgraf** [darribas/contextily#9]
Hey there - I'm working to convert some geomapping code from R into Python, and looking to recreate some of the `get_map` functionality. It seems like this package along with `geopy` or some other geocoding tool could make it easy to return a map corresponding to a search query. Have you thought about adding in this feature?

It doesn't seem like it'd be *too* tough and could be quite useful. Imagine just doing `ctx.get_map('texas')` and it'd try to be smart about the bounding box size, zoom level, etc.

**Comment by darribas** [comment]
Hi @choldgraf! Thanks for chipping in the idea. I think it's great and would enhance the library. I hadn't thought about it but I did consider at some point using some sort of "smartness" to pick a default zoom level, although never got around to implement it (instead, I just wrote [`howmany`](https://github.com/darribas/contextily/blob/master/contextily/__init__.py#L149)).

At the moment, I'm pretty low on bandwith so don't have much time to work on `contextily`, but I'd definitely welcome PRs in this direction and would be happy to review them.

**Comment by choldgraf** [comment]
Lemme see what I throw together for this class, and if it's something worth contributing upstream I'll give it a go :-)

**Comment by knaaptime** [comment]
I'd be happy to help with this where I can

the geopy Nominatim geocoder returns a bounding box as part of the `raw` attribute, so the first half of `get_map` should be pretty straightforward. More difficult to figure out sensible default zoom to pass to `howmany`

```

In [1]: from geopy.geocoders import Nominatim

In [2]: balt = Nominatim().geocode('baltimore')

In [3]: balt.raw['boundingbox']
Out[3]: ['39.1308816', '39.4508816', '-76.770759', '-76.450759']

```

**Comment by choldgraf** [comment]
Yep - this is basically why I brought up the issue! It feels like the individual components are there. Trying to figure out if this is truly low-hanging fruit or if there is a dragon in there somewhere ;-)

**Comment by knaaptime** [comment]
oh cool. I hadn't realized geopy returned a bbox until just now. id be happy to see this implemented because i use contextily a lot

**Comment by choldgraf** [comment]
lemme give it a shot and see how it works!

**Comment by choldgraf** [comment]
Here's a prototype:

```
class Location(object):
    
    def __init__(self, place, zoom=5, max_tiles=20, latlon=True, file=None):
        self.place = place
        self.zoom = zoom
        self.max_tiles = max_tiles
        self.latlon = latlon
        self.file = file

        # Get geocoded values
        resp = gp.geocoders.Nominatim().geocode(place)
        bbox = np.array([float(ii) for ii in resp.raw['boundingbox']])
        
        self.geocode = resp
        self.s, self.n, self.w, self.e = bbox
        im, bbox = self.get_map()

    def get_map(self):
        n_tiles = cx.howmany(self.w, self.s, self.e, self.n, self.zoom, ll=self.latlon)
        if n_tiles > self.max_tiles:
            raise ValueError('Too many tiles downloaded ({})'.format(n_tiles))
        
        if isinstance(self.file, str):
            im, bbox = cx.bounds2raster(self.w, self.s, self.e, self.n, self.zoom, ll=self.latlon)
        else:
            im, bbox = cx.bounds2img(self.w, self.s, self.e, self.n, self.zoom, ll=self.latlon)
        
        self.im = im
        self.bbox = bbox
        return im, bbox
    
    def plot_map(self, ax=None, axis_off=True):
        
        if not hasattr(self, 'im'):
            raise ValueError('Call `get_map` before plotting a map')
        
        if ax is None:
            fig, ax = plt.subplots(figsize=(10, 10))
        ax.imshow(self.im, extent=self.bbox)
        ax.set(title=self.place, xlabel="Longitude", ylabel="Latitude")
        if axis_off is True:
            ax.set_axis_off()
        return ax
    
def function_to_choose_zoom_based_on_bbox?????
   ...
```

API call is like this:

```
loc = Location('uc berkeley', zoom=14)
loc.plot_map()
```

And this plots:

![image](https://user-images.githubusercontent.com/1839645/26846485-a7ae2962-4aae-11e7-9525-688e068e89ec.png)

Thoughts on the API? Right now I'm handling it as a class because some of the `N/S/E/W` stuff with the bounding box is a little confusing. It could be handled by a few functions too though. Curious to see what you think!

**Comment by lwasser** [comment]
this sounds awesome! i'm going to be teaching with this next year so thank you everyone for thinking about this 👍 and @choldgraf for all of the hard work!!

**Comment by darribas** [comment]
@choldgraf This looks like a great start!!! A couple of thoughts:

1. I'm not too concerned about class/method in general. In theory, I'd tend to favor methods as they're simpler to read and maintain but if there are good reasons to use classes, I'm all in.
1. The `plot_map()` method looks like something that'd be pretty useful not only within `Location`. Any scope to pull it apart?

It'd be great to get a PR with tests included, I agree with @lwasser, students (and my own research) would love this 😄

**Comment by choldgraf** [comment]
How would you suggest writing tests for something like this? The main class will pull something from the web, and I'm not sure we'd want that dependency since sometimes (e.g., now) the tile API is down.

**Comment by darribas** [comment]
Mmm... good point. My view would be to make everything as modular as possible and test each component separately. In the case of what needs to be pulled down, we could have a downloaded file that is loaded for the rest of the tests. In any case, the API being down is something we probably want to catch anyway, so might be useful to test that too (that's why I wrote the simpler tests for the [tile providers we support](https://github.com/darribas/contextily/blob/master/tests/test_providers.py))?

For examples of more standard tests, have a look at the library's [current ones](https://github.com/darribas/contextily/blob/master/tests/test_ctx.py).
